plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go

```diff
@@ -87,24 +113,62 @@ func (s *SelectorSpread) CalculateSpreadPriority(pod *api.Pod, podLister algorit
 				}
 			}
 			if matches {
-				counts[pod.Spec.NodeName]++
+				countsByNodeName[pod.Spec.NodeName]++
 				// Compute the maximum number of pods hosted on any node
-				if counts[pod.Spec.NodeName] > maxCount {
-					maxCount = counts[pod.Spec.NodeName]
+				if countsByNodeName[pod.Spec.NodeName] > maxCountByNodeName {
+					maxCountByNodeName = countsByNodeName[pod.Spec.NodeName]
 				}
 			}
 		}
 	}
 
+	maxCountByZone := 0
+	haveZones := false
+	countsByZone := map[string]int{}
+	for i := range nodes.Items {
```

Comment:
The first iteration was over the pods; now we are iterating over the nodes to find the counts-by-zone.  We could build a map of node-name to node and then combine these two loops, but we'd need a loop to build that map (and I think this is more efficient than building that node-map).

I am making this a little more efficient and perhaps easier to understand though, by computing the aggregates in a separate loop rather than in the body of the main loop.  It does mean more loops, but each loop is much simpler.